File: src/trading/quoter.py
"""
src/trading/quoter.py

Minimal two-sided quoting loop — the first code that places orders which CAN
fill and lose real money. Purpose: MEASURE the real adverse-selection bleed and
best-price presence of liquidity provision in ONE incentive pool, at tiny size,
before scaling capital or porting to always-on infra.

SAFETY (real money; every knob small, every exit flattens):
  • size/position: quotes `size` contracts/side (default 2), inventory capped at
    ±max_position (default 4). In a $0.10–0.50 market that bounds absolute risk
    to a few dollars even in a total wipeout.
  • bounded runtime: exits after --minutes.
  • max-loss kill: if marked PnL ≤ −max_loss, cancel + flatten + stop.
  • ALWAYS flattens on EVERY exit path (timeout, error, Ctrl-C) via try/finally:
    cancels resting orders AND closes open inventory with a marketable order
    priced through the touch, then verifies flat and SCREAMS if not.
  • only quotes a two-sided book (the subsidy's own requirement anyway).
  • self-tracks position from ITS OWN fills (starts flat, only its orders move
    it), so the cap and flatten never depend on guessing API position fields.

Strategy: join best bid and best ask (be AT best for the 1.0x incentive score),
re-quote when the touch moves, stop quoting a side once it would push inventory
past the cap. Logs each cycle so fills and bleed are visible live.

Usage (REAL MONEY — needs --yes):
    ./.venv/bin/python -m src.trading.quoter --ticker KXWCADS-26JUL19-ABNB \
        --size 2 --max-position 4 --minutes 15 --yes
"""

from __future__ import annotations

import argparse
import time
from decimal import Decimal

from src.pipeline.auth import CANDIDATE_HOSTS, KalshiClient
from src.trading import orders

PROD_HOST = CANDIDATE_HOSTS[0]
MAKER_RATE = Decimal("0.0175")
MIN_P, MAX_P = Decimal("0.01"), Decimal("0.99")


def maker_fee(p: Decimal, n: int) -> Decimal:
    return MAKER_RATE * p * (1 - p) * n


def _remaining(o: dict) -> Decimal:
    r = o.get("remaining_count") or o.get("remaining_count_fp")
    if r is not None:
        return Decimal(str(r))
    init = Decimal(str(o.get("initial_count_fp", o.get("initial_count", "0"))))
    fill = Decimal(str(o.get("fill_count_fp", o.get("fill_count", "0"))))
    return init - fill


class Quoter:
    def __init__(self, client, ticker, size, max_position, max_loss):
        self.c = client
        self.ticker = ticker
        self.size = size
        self.max_position = max_position
        self.max_loss = Decimal(str(max_loss))
        self.pos = 0                 # net YES contracts (self-tracked from fills)
        self.cash = Decimal(0)       # realized cash from fills
        self.fees = Decimal(0)
        self.fills = 0
        self.bid = None              # {"id","price","rem"} or None
        self.ask = None
        self.cycles = 0
        self.at_best = 0             # cycles resting two-sided at best

    # --- pnl ---------------------------------------------------------------
    def pnl(self, mid: Decimal) -> Decimal:
        return self.cash + Decimal(self.pos) * mid - self.fees

    def _record_fill(self, side: str, price: Decimal, n: Decimal) -> None:
        n_i = int(n)
        if n_i <= 0:
            return
        if side == "bid":                       # bought YES at my resting bid
            self.pos += n_i
            self.cash -= price * n_i
        else:                                   # sold YES at my resting ask
            self.pos -= n_i
            self.cash += price * n_i
        self.fees += maker_fee(price, n_i)
        self.fills += 1
        print(f"    FILL {side} {n_i}@{price}  -> pos={self.pos} cash={self.cash:.4f}")

    # --- fill detection: diff my tracked orders vs the live resting list ----
    def _detect_fills(self) -> None:
        live = {o.get("order_id") or o.get("id"): o for o in orders.resting_orders(self.c)}
        for tag in ("bid", "ask"):
            o = getattr(self, tag)
            if o is None:
                continue
            if o["id"] in live:
                rem = _remaining(live[o["id"]])
                if rem < o["rem"]:
                    self._record_fill(tag, o["price"], o["rem"] - rem)
                    o["rem"] = rem
                if rem <= 0:
                    setattr(self, tag, None)
            else:  # gone and we didn't cancel it -> fully filled (GTC won't expire)
                self._record_fill(tag, o["price"], o["rem"])
                setattr(self, tag, None)

    def _requote_side(self, tag: str, want: bool, price: Decimal | None) -> None:
        o = getattr(self, tag)
        if not want or price is None:
            if o is not None:
                orders.cancel(self.c, o["id"]); setattr(self, tag, None)
            return
        if o is not None and o["price"] == price:
            return  # already at best on this side — leave it (keep queue priority)
        if o is not None:
            orders.cancel(self.c, o["id"]); setattr(self, tag, None)
        try:
            placed = orders.rest_limit(self.c, self.ticker, tag, price,
                                       count=self.size, post_only=True, guard=False)
            oid = placed.get("order_id") or placed.get("id")
            setattr(self, tag, {"id": oid, "price": price, "rem": Decimal(self.size)})
        except orders.OrderError as e:
            print(f"    (skip {tag} @ {price}: {e})")  # post_only rejected a racing quote

    # --- one cycle ---------------------------------------------------------
    def step(self) -> bool:
        """Returns False to stop (kill switch)."""
        self.cycles += 1
        self._detect_fills()
        bid, ask = orders._best_quote(self.c, self.ticker)
        if bid is None or ask is None:
            return True  # one-sided book: don't quote, just wait
        mid = (bid + ask) / 2
        pnl = self.pnl(mid)
        two_sided = self.bid is not None and self.ask is not None \
            and self.bid["price"] == bid and self.ask["price"] == ask
        if two_sided:
            self.at_best += 1
        print(f"  [{self.cycles:>3}] bid {bid} ask {ask} | pos {self.pos} "
              f"pnl {pnl:+.4f} fills {self.fills} atbest {self.at_best}")
        if pnl <= -self.max_loss:
            print(f"  !! KILL: pnl {pnl:.4f} ≤ -{self.max_loss}")
            return False
        self._requote_side("bid", self.pos < self.max_position, bid)
        self._requote_side("ask", self.pos > -self.max_position, ask)
        return True

    # --- flatten everything (called on EVERY exit) -------------------------
    def flatten(self) -> None:
        print("flattening: cancelling all resting orders ...")
        try:
            orders.cancel_all(self.c)
        except Exception as e:  # noqa: BLE001
            print(f"  cancel_all error: {e}")
        self.bid = self.ask = None
        if self.pos != 0:
            bid, ask = orders._best_quote(self.c, self.ticker)
            # price THROUGH the touch to guarantee the close fills (a few ¢ slip
            # on ≤max_position contracts — being flat beats saving pennies).
            if self.pos > 0 and bid is not None:      # long -> sell YES
                px = max(bid - Decimal("0.03"), MIN_P)
                side = "ask"
            elif self.pos < 0 and ask is not None:    # short -> buy YES
                px = min(ask + Decimal("0.03"), MAX_P)
                side = "bid"
            else:
                print(f"  !! cannot price flatten (one-sided book); pos={self.pos} LEFT OPEN")
                return
            print(f"  closing pos={self.pos} via marketable {side} {abs(self.pos)}@{px}")
            try:
                orders.rest_limit(self.c, self.ticker, side, px, count=abs(self.pos),
                                  post_only=False, guard=False)
            except orders.OrderError as e:
                print(f"  !! flatten order FAILED: {e}")
        time.sleep(1.0)
        left = orders.resting_orders(self.c)
        mp = orders.positions(self.c).get("market_positions", [])
        openpos = [p for p in mp if p.get("ticker") == self.ticker]
        if left or openpos:
            print(f"  !! NOT FLAT — resting={len(left)} positions={openpos} — CHECK MANUALLY")
        else:
            print("  verified flat: 0 resting, 0 positions.")

    def summary(self) -> None:
        bid, ask = orders._best_quote(self.c, self.ticker)
        mid = (bid + ask) / 2 if (bid and ask) else Decimal("0.5")
        print(f"\n{'='*56}\nSESSION SUMMARY  {self.ticker}\n{'='*56}")
        print(f"  cycles {self.cycles} | two-sided-at-best {self.at_best} "
              f"({self.at_best/max(self.cycles,1):.0%})")
        print(f"  fills {self.fills} | final pos {self.pos}")
        print(f"  realized cash {self.cash:+.4f} | fees {self.fees:.4f} "
              f"| est PnL {self.pnl(mid):+.4f}")
        print("  (bleed = PnL after flatten; positive best-price% = subsidy-qualifying)\n")


def run(client, ticker, size, max_position, minutes, max_loss, poll) -> None:
    # precondition: start FLAT. Never begin a quoting session with open risk.
    if orders.resting_orders(client):
        raise SystemExit("account has resting orders — cancel-all first.")
    if [p for p in orders.positions(client).get("market_positions", []) if p.get("ticker") == ticker]:
        raise SystemExit(f"already hold a position in {ticker} — flatten first.")

    q = Quoter(client, ticker, size, max_position, max_loss)
    deadline = time.monotonic() + minutes * 60
    print(f"quoting {ticker}  size={size} maxpos=±{max_position} "
          f"maxloss=${max_loss} for {minutes}min\n")
    try:
        while time.monotonic() < deadline:
            if not q.step():
                break
            time.sleep(poll)
    except KeyboardInterrupt:
        print("\ninterrupted.")
    finally:
        q.flatten()      # the load-bearing safety net — runs no matter what
        q.summary()


def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--ticker", required=True)
    ap.add_argument("--size", type=int, default=2, help="contracts per side")
    ap.add_argument("--max-position", type=int, default=4)
    ap.add_argument("--minutes", type=float, default=15)
    ap.add_argument("--max-loss", type=float, default=5.0, help="dollars")
    ap.add_argument("--poll", type=float, default=2.0, help="seconds between cycles")
    ap.add_argument("--yes", action="store_true", help="required — places REAL orders")
    args = ap.parse_args()
    if not args.yes:
        raise SystemExit("This places REAL, fillable orders. Re-run with --yes.")
    client = KalshiClient.from_env(host=PROD_HOST)
    run(client, args.ticker, args.size, args.max_position,
        args.minutes, args.max_loss, args.poll)


if __name__ == "__main__":
    main()

File: cf-quoter/src/quoter.ts
/**
 * src/quoter.ts — QuoterDO: an always-on two-sided quoter for ONE market.
 *
 * Port of the proven Python quoter.py, restructured for a Durable Object:
 * instead of a sleep loop, each quoting cycle runs in alarm(), which reschedules
 * itself every pollSeconds. All state is persisted to DO storage, so an eviction
 * or crash resumes cleanly (the exchange remembers our resting orders; we reload
 * our tracking of them).
 *
 * REAL MONEY, UNATTENDED. Safety (mirrors quoter.py, and matters MORE here
 * because nothing else is watching):
 *   • size/maxPosition caps bound exposure to a few dollars.
 *   • deadline (endTime): stops + flattens after the configured window.
 *   • max-loss kill: marked PnL ≤ −maxLoss → flatten + stop.
 *   • stop()/deadline/kill ALL flatten: cancel resting orders AND close any
 *     inventory with a marketable order priced through the touch.
 *   • start() refuses to begin unless the account is flat (no resting orders,
 *     no position) — never resume on top of unknown risk.
 *   • self-tracks position from its own fills, persisting last-seen remaining so
 *     an alarm retry can't double-count.
 */

import { DurableObject } from "cloudflare:workers";
import { Kalshi, fp } from "./kalshi";
import type { Env } from "./index";

const MAKER_RATE = 0.0175;
// v2 position-skew: once |inventory| reaches this fraction of maxPosition, start
// improving the REDUCING side by one tick so the position works back to flat
// instead of pinning at the cap (§ v2 learnings — the USPPIYOY pin).
const SKEW_SOFT_FRAC = 0.6;
const TICK = 0.01;

interface Leg { id: string; price: string; rem: number; }
interface State {
  ticker: string;
  size: number;
  maxPosition: number;
  maxLoss: number;       // dollars
  pollSeconds: number;
  endTime: number;       // epoch ms
  running: boolean;
  stopReason: string;
  position: number;
  cash: number;
  fees: number;
  fills: number;
  cycles: number;
  atBest: number;
  bid: Leg | null;
  ask: Leg | null;
}

export interface StartConfig {
  ticker: string;
  size?: number;
  maxPosition?: number;
  maxLoss?: number;
  pollSeconds?: number;
  minutes?: number;
}

function remaining(o: any): number {
  const r = o.remaining_count ?? o.remaining_count_fp;
  if (r != null) return Number(r);
  return Number(o.initial_count_fp ?? o.initial_count ?? 0)
    - Number(o.fill_count_fp ?? o.fill_count ?? 0);
}

export class QuoterDO extends DurableObject<Env> {
  private k?: Kalshi;

  private async kalshi(): Promise<Kalshi> {
    if (!this.k) {
      this.k = await Kalshi.create(this.env.KALSHI_API_KEY_ID, this.env.KALSHI_PRIVATE_KEY_B64);
    }
    return this.k;
  }

  // Resting orders for THIS DO's market only. The exchange returns every order
  // on the account, but with one DO per ticker we must never touch another
  // market's orders — the whole point of per-ticker scoping. Every place that
  // cancels or flat-checks goes through this filter.
  private async myOrders(k: Kalshi, ticker: string): Promise<any[]> {
    return (await k.restingOrders()).filter(
      (o) => (o.ticker ?? o.market_ticker) === ticker,
    );
  }

  // --- control (RPC) ------------------------------------------------------
  async start(cfg: StartConfig): Promise<object> {
    const k = await this.kalshi();
    // Flat-check is PER TICKER now: another market's resting orders are none of
    // this DO's business (the old account-wide check made simultaneous markets
    // impossible — the second start() always saw the first market's orders).
    if ((await this.myOrders(k, cfg.ticker)).length) throw new Error(`resting orders already in ${cfg.ticker}`);
    // Kalshi returns the signed contract count as `position_fp` (a string) — NOT
    // `position` (which doesn't exist, so reading it silently sees 0 and would let
    // us start ON TOP of an open position). This is the field bug that left run-1
    // econ positions orphaned; the flat-check must read position_fp.
    const pos = await k.positions(cfg.ticker);
    if (pos.some((p) => Number(p.position_fp ?? p.position ?? 0) !== 0)) throw new Error(`open position in ${cfg.ticker}`);

    // Buying-power guard: the $147 account is SHARED across every market, so a
    // new market must not commit more than the account can currently cover.
    // Worst case this market holds a full one-sided position of `maxPosition`
    // contracts (~$1 each), so require at least that much free balance now.
    const need = cfg.maxPosition ?? 4;
    const bal = Number(await k.balanceDollars());
    if (bal < need) throw new Error(`balance $${bal.toFixed(2)} < ~$${need} needed for maxPosition ${need}`);

    const minutes = cfg.minutes ?? 180;
    const s: State = {
      ticker: cfg.ticker,
      size: cfg.size ?? 2,
      maxPosition: cfg.maxPosition ?? 4,
      maxLoss: cfg.maxLoss ?? 5,
      pollSeconds: cfg.pollSeconds ?? 10,
      endTime: Date.now() + minutes * 60_000,
      running: true, stopReason: "",
      position: 0, cash: 0, fees: 0, fills: 0, cycles: 0, atBest: 0,
      bid: null, ask: null,
    };
    await this.ctx.storage.put("state", s);
    await this.ctx.storage.setAlarm(Date.now() + 1000); // first cycle shortly
    return { started: true, ...this.snapshot(s) };
  }

  async stop(): Promise<object> {
    const s = await this.ctx.storage.get<State>("state");
    if (!s) return { stopped: true, note: "no active session" };
    await this.doStop(s, "manual");
    return { stopped: true, ...this.snapshot(s) };
  }

  async status(): Promise<object> {
    const s = await this.ctx.storage.get<State>("state");
    return s ? this.snapshot(s) : { running: false, note: "never started" };
  }

  private snapshot(s: State) {
    return {
      ticker: s.ticker, running: s.running, stopReason: s.stopReason,
      position: s.position, fills: s.fills, cycles: s.cycles, atBest: s.atBest,
      atBestPct: s.cycles ? Math.round((100 * s.atBest) / s.cycles) : 0,
      cash: +s.cash.toFixed(4), fees: +s.fees.toFixed(4),
      endsInMin: Math.max(0, Math.round((s.endTime - Date.now()) / 60_000)),
    };
  }

  // --- the loop -----------------------------------------------------------
  async alarm(): Promise<void> {
    const s = await this.ctx.storage.get<State>("state");
    if (!s || !s.running) return;
    const k = await this.kalshi();
    try {
      if (Date.now() >= s.endTime) { await this.doStop(s, "deadline"); return; }
      await this.detectFills(k, s);
      const q = await k.bestQuote(s.ticker);
      if (q.bid && q.ask) {
        const mid = (Number(q.bid) + Number(q.ask)) / 2;
        const pnl = s.cash + s.position * mid - s.fees;
        s.cycles++;
        if (s.bid && s.ask && fp(s.bid.price) === fp(q.bid) && fp(s.ask.price) === fp(q.ask)) s.atBest++;
        if (pnl <= -s.maxLoss) { await this.doStop(s, `killswitch pnl=${pnl.toFixed(2)}`); return; }
        // Position-skew the PRICES. Below the soft threshold both sides sit at
        // best (symmetric — capture spread + subsidy). Past it, improve the
        // reducing side by one tick to become the new best there, so it gets
        // filled and unwinds inventory (a resting quote AT best didn't fill and
        // pinned us last run). The one-tick improve is clamped so it can never
        // cross the book (post_only would reject a cross anyway), so on a 1¢
        // market it simply stays at best. Maker-only — no taker fee.
        let bidPx = Number(q.bid), askPx = Number(q.ask);
        if (Math.abs(s.position) >= s.maxPosition * SKEW_SOFT_FRAC) {
          if (s.position > 0) askPx = Math.max(bidPx + TICK, askPx - TICK);      // long → cheaper ask, sell down
          else if (s.position < 0) bidPx = Math.min(askPx - TICK, bidPx + TICK); // short → richer bid, buy back
        }
        // Pass the ROOM on each side (contracts until the ±cap), so requote can
        // size the order down and a full fill can never breach maxPosition.
        await this.requote(k, s, "bid", s.maxPosition - s.position, fp(bidPx));
        await this.requote(k, s, "ask", s.maxPosition + s.position, fp(askPx));
      }
      await this.ctx.storage.put("state", s);
      await this.ctx.storage.setAlarm(Date.now() + s.pollSeconds * 1000);
    } catch (e) {
      // transient (network/rate-limit): persist + reschedule so the loop
      // survives a blip. Placing no order on error is safe; fills reconcile
      // next cycle from the exchange's own resting-order list.
      console.log("alarm error:", (e as Error).message);
      await this.ctx.storage.put("state", s);
      await this.ctx.storage.setAlarm(Date.now() + s.pollSeconds * 1000);
    }
  }

  private async detectFills(k: Kalshi, s: State): Promise<void> {
    const live = new Map<string, any>();
    for (const o of await k.restingOrders()) live.set(o.order_id ?? o.id, o);
    for (const tag of ["bid", "ask"] as const) {
      const o = s[tag];
      if (!o) continue;
      if (live.has(o.id)) {
        const rem = remaining(live.get(o.id));
        if (rem < o.rem) { this.recordFill(s, tag, o.price, o.rem - rem); o.rem = rem; }
        if (rem <= 0) s[tag] = null;
      } else {
        this.recordFill(s, tag, o.price, o.rem); // gone & we didn't cancel -> filled
        s[tag] = null;
      }
    }
  }

  private recordFill(s: State, tag: "bid" | "ask", price: string, n: number): void {
    if (n <= 0) return;
    const p = Number(price);
    if (tag === "bid") { s.position += n; s.cash -= p * n; }
    else { s.position -= n; s.cash += p * n; }
    s.fees += MAKER_RATE * p * (1 - p) * n;
    s.fills++;
    console.log(`FILL ${tag} ${n}@${price} -> pos=${s.position} cash=${s.cash.toFixed(4)}`);
  }

  // `room` = contracts allowed on this side before hitting the ±maxPosition cap.
  // We quote min(size, room), so even a full fill can't breach the cap — this is
  // the fix for the 53-vs-40 overshoot (the old code checked the cap but still
  // placed a full-size order, so effective cap was maxPosition + size).
  private async requote(k: Kalshi, s: State, tag: "bid" | "ask", room: number, price: string): Promise<void> {
    const want = Math.min(s.size, Math.max(0, room));
    const o = s[tag];
    if (want <= 0) { if (o) { await k.cancel(o.id); s[tag] = null; } return; }
    if (o && fp(o.price) === fp(price) && o.rem <= room) return; // at best & can't breach — keep it
    if (o) { await k.cancel(o.id); s[tag] = null; }
    try {
      const id = await k.restLimit(s.ticker, tag, fp(price), want, true);
      s[tag] = { id, price: fp(price), rem: want };
    } catch (e) {
      console.log(`skip ${tag}@${price}: ${(e as Error).message}`); // post_only rejected a racing quote
    }
  }

  // --- stop + flatten (runs on manual stop, deadline, and kill) -----------
  private async doStop(s: State, reason: string): Promise<void> {
    const k = await this.kalshi();
    s.running = false; s.stopReason = reason;
    console.log(`STOP (${reason}) — flattening`);
    try {
      // Cancel ONLY this market's resting orders. The old code cancelled every
      // order on the account, so one market stopping would wipe out every OTHER
      // market's quotes — the cross-DO bug that made multi-market unsafe.
      for (const o of await this.myOrders(k, s.ticker)) await k.cancel(o.order_id ?? o.id).catch(() => {});
      s.bid = s.ask = null;
      if (s.position !== 0) {
        const q = await k.bestQuote(s.ticker);
        let side: "bid" | "ask" | null = null, px = "";
        if (s.position > 0 && q.bid) { side = "ask"; px = fp(Math.max(Number(q.bid) - 0.03, 0.01)); }
        else if (s.position < 0 && q.ask) { side = "bid"; px = fp(Math.min(Number(q.ask) + 0.03, 0.99)); }
        if (side) {
          console.log(`closing pos=${s.position} via marketable ${side} ${Math.abs(s.position)}@${px}`);
          await k.restLimit(s.ticker, side, px, Math.abs(s.position), false).catch(
            (e) => console.log("!! flatten order FAILED:", (e as Error).message));
        } else {
          console.log(`!! cannot price flatten (one-sided book); pos=${s.position} LEFT OPEN`);
        }
      }
    } finally {
      await this.ctx.storage.deleteAlarm();
      await this.ctx.storage.put("state", s);
      // Drop ourselves from the live-market registry so a self-stop (deadline or
      // kill-switch) doesn't leave a zombie entry that /status keeps polling.
      await this.env.REGISTRY.getByName("registry").remove(s.ticker).catch(() => {});
    }
  }
}
